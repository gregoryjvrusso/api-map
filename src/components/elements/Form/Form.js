import React from 'react'
import axios from 'axios'
import Label from '../Label/Label'
import InputCEP from '../InputCEP/InputCEP'
import Button from '../Button/Button'

const Form = ({ setCep, setData, setMessage, cep }) => {
  const fetchData = async (url) => {
    try {
      const result = await axios(url)
      if (result.data.erro) {
        setData('')
        setMessage(true)
      } else {
        setData(result.data)
        setMessage(false)
      }
    } catch (error) {
      console.log(error)
    }
  }

  const validatorCep = (cep) => {
    const regex = RegExp(/^\d{5}-\d{3}$/)
    if (regex.test(cep)) {
      let cepCheck = cep.replace('-', '')
      fetchData(`https://viacep.com.br/ws/${cepCheck}/json/`)
    } else {
      setMessage(true)
      setData('')
    }
  }

  return (
    <form onSubmit={e => {
      e.preventDefault()
      validatorCep(cep)
    }}>
      <Label htmlFor='cep'>CEP</Label>
      <InputCEP
        width={['70%', '15em', '15em']}
        name='cep'
        value={cep}
        onChange={(e) => {
          setMessage(false)
          setCep(e.target.value)
        }}
        id='cep'
        type='text'
        placeholder='02250-250'
      />
      <Button 
        width={['95%', '8em', '8em']} 
        type='submit' 
      >
        Buscar
            </Button>
    </form>
  )
}

export default Form