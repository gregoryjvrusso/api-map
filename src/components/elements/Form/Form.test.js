import React, { useState } from 'react'
import { render, fireEvent, waitForElement } from '@testing-library/react'
import Form from './Form.js'

const ComponentMock = () => {

  const validatorCep = (cep) => {
    const regex = RegExp(/^\d{5}-\d{3}$/)
    if (regex.test(cep)) {
      let cepCheck = cep.replace('-', '')
      setData(cepCheck)
    } else {
      setData('CEP inválido')
    }
  }

  const [cep, setCep] = useState('')
  const [data, setData] = useState(false)
  return (
    <React.Fragment>
      <p data-testid='return-form'>{data}</p>
      <form onSubmit={e => {
        e.preventDefault()
        validatorCep(cep)
      }}>
        <label htmlFor='cep'>CEP</label>
        <input
          width={['70%', '15em', '15em']}
          name='cep'
          value={cep}
          onChange={(e) => {
            setCep(e.target.value)
          }}
          id='cep'
          type='text'
          data-testid='input-form'
        />
        <button
          type='submit'
          data-testid='button-form'
        >
          Buscar
      </button>
      </form>
    </React.Fragment>
  )
}

describe('Form component', () => {
  it('deve conseguir renderizar [snapshot]', async () => {
    const { container } = render(<Form />)
    expect(container.firstChild).toMatchSnapshot()

  })
  it('deve conseguir editar o input', async () => {
    const { getByTestId } = render(<ComponentMock />)
    const input = getByTestId('input-form')
    fireEvent.change(input, { target: { value: '11704-690' } })
    expect(input.value).toBe('11704-690')
  })
  it('deve validar o CEP e retornar', async () => {
    const { getByTestId, getByText } = render(<ComponentMock />)
    const input = getByTestId('input-form')
    fireEvent.change(input, { target: { value: '11704-690' } })
    const button = getByTestId('button-form')
    fireEvent.click(button)    
    await waitForElement(() => getByText('11704690'))
  })
  it('deve retornar erro de validação de CEP por falta de número', async () => {
    const { getByTestId, getByText } = render(<ComponentMock />)
    const input = getByTestId('input-form')
    fireEvent.change(input, { target: { value: '11704' } })
    const button = getByTestId('button-form')
    fireEvent.click(button)    
    await waitForElement(() => getByText('CEP inválido'))
  })
  it('deve retornar erro de validação de CEP por passar letra ao invés de número', async () => {
    const { getByTestId, getByText } = render(<ComponentMock />)
    const input = getByTestId('input-form')
    fireEvent.change(input, { target: { value: '11704-AHD' } })
    const button = getByTestId('button-form')
    fireEvent.click(button)    
    await waitForElement(() => getByText('CEP inválido'))
  })
})
