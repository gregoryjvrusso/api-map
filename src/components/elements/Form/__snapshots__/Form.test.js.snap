// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component deve conseguir renderizar [snapshot] 1`] = `
<form>
  <label
    class="sc-bdVaJa FTQhB"
    for="cep"
  >
    CEP
  </label>
  <input
    class="sc-bwzfXH fRVSPy"
    data-testid="input-form"
    id="cep"
    name="cep"
    placeholder="02250-250"
    type="text"
    value=""
    width="70%,15em,15em"
  />
  <button
    class="sc-htpNat irRmst"
    type="submit"
    width="95%,8em,8em"
  >
    Buscar
  </button>
</form>
`;
